guides/source/association_basics.md

```diff
@@ -594,1804 +784,1332 @@ class Book < ApplicationRecord
+    irb> author.books.any? do |book|
+    irb>   book.writer.equal?(author) # This executes an author query for every book
+    irb> end
+    => false
+    ```
 
-class Representative < ApplicationRecord
-  has_many :accounts
-end
-```
+* Reference multiple copies of a model with inconsistent data:
 
-##### `readonly`
+    ```irb
+    irb> author = Author.first
+    irb> book = author.books.first
+    irb> author.name == book.writer.name
+    => true
+    irb> author.name = "Changed Name"
+    irb> author.name == book.writer.name
+    => false
+    ```
 
-If you use the `readonly` method, then the associated object will be read-only when retrieved via the association.
+* Fail to autosave associations:
 
-##### `select`
+    ```irb
+    irb> author = Author.new
+    irb> book = author.books.new
+    irb> book.save!
+    irb> book.persisted?
+    => true
+    irb> author.persisted?
+    => false
+    ```
 
-The `select` method lets you override the SQL `SELECT` clause that is used to retrieve data about the associated object. By default, Rails retrieves all columns.
+* Fail to validate presence or absence:
 
-#### Do Any Associated Objects Exist?
+    ```irb
+    irb> author = Author.new
+    irb> book = author.books.new
+    irb> book.valid?
+    => false
+    irb> book.errors.full_messages
+    => ["Author must exist"]
+    ```
 
-You can see if any associated objects exist by using the `association.nil?` method:
+To resolve these issues, you can explicitly declare bi-directional associations using the `:inverse_of` option:
 
 ```ruby
-if @supplier.account.nil?
-  @msg = "No account found for this supplier"
+class Author < ApplicationRecord
+  has_many :books, inverse_of: 'writer'
 end
-```
 
-#### When are Objects Saved?
+class Book < ApplicationRecord
+  belongs_to :writer, class_name: 'Author', foreign_key: 'author_id'
+end
+```
 
-When you assign an object to a `has_one` association, that object is automatically saved (in order to update its foreign key). In addition, any object being replaced is also automatically saved, because its foreign key will change too.
+By including the `:inverse_of` option in the `has_many` association declaration,
+Active Record will recognize the bi-directional association and behave as described in
+the initial examples above.
 
-If either of these saves fails due to validation errors, then the assignment statement returns `false` and the assignment itself is cancelled.
+[`config.active_record.automatic_scope_inversing`]: configuring.html#config-active-record-automatic-scope-inversing
 
-If the parent object (the one declaring the `has_one` association) is unsaved (that is, `new_record?` returns `true`) then the child objects are not saved. They will automatically when the parent object is saved.
+Detailed Association References
+------------------------------
 
-If you want to assign an object to a `has_one` association without saving the object, use the `build_association` method.
+### References
```

Comment:
Just to double check, did you mean for us to take a look at just this 'References' section, or the entire parent 'Detailed Association Reference' parent?

I think it's a tricky one to get right. Gut feel is that there's a lot of (maybe unnecessary) information here, but I think that's just because it almost serves as a quick glossary of methods that people will refer to. It's also repetitive, and I've wondered about formatting slightly differently to sort that, but again, if you're coming here to find out about the available methods for `has_one`, for example, you probably want them all in one place.

I think it's clear though, and serves a purpose. Maybe let's see what the community come back with when we put the changes out to them - if they think it's unwieldy, then maybe there's a way of splitting this section out to somewhere else.

Reply:
> Just to double check, did you mean for us to take a look at just this 'References' section, or the entire parent 'Detailed Association Reference' parent?

Mostly the "References" section although I've now pulled "Scopes" and "options" out into their own sections, so I'm referring to those three sections. 

> Gut feel is that there's a lot of (maybe unnecessary) information here

I feel very similarly and I'm concerned that it's not super useful and like you say repetitive (although we've done a better job with the examples here in the guides compared to the API) 

> Maybe let's see what the community come back with when we put the changes out to them - if they think it's unwieldy, then maybe there's a way of splitting this section out to somewhere else.

This sounds like a sane way forward - lets see if Petrik has any immediate suggestions and if not we'll get community feedback.

Reply:
I'm also not sure about the "References" section. It reads more like the API documentation, listing a lot of methods, instead of a guide (this PR improves this a lot already). I'm in favor of removing as much as possible from this section. On the other hand associations are a heavily used Rails API, so it deserves more than just linking to the API documentation.

The descriptions of the added methods are sort of hidden in the "References" section now.
Maybe we can merge the methods for each association into the "Types of Associations" sections?
This makes sure each association is described in a single section.
As the associations are such an important API, maybe each of the six associations can be a top section?
This would also clarify what each section contains in the sidebar.

The methods added for `has_many`, `has_and_belongs_to_many` and `has_many :through` are mostly identical.
So I think in the "Reference" section we only need to mention the methods for `has_many` and mention they are similar for  `has_and_belongs_to_many` and `has_many :through` (this one already seems deleted in the PR).

I hope these suggestions are somewhat helpful. Otherwise, let's get this out to the community.

Reply:
@p8 these suggestions were really helpful, however, there are parts that I'm not sure that I'm following. Here's my thoughts:

> Maybe we can merge the methods for each association into the "Types of Associations" sections?
This makes sure each association is described in a single section.

This makes sense to me, and cleans it up nicely!

> As the associations are such an important API, maybe each of the six associations can be a top section?

I think that you're suggesting that we should move this out of the "Types of Associations" section and make them level-1 headers. If this is correct, I'm not sure that I feel aligned with this, as we will only show one subheading "Methods added by .." under each section. I like that we have the heading "Types of Associations" to group them. 

> The methods added for has_many, has_and_belongs_to_many and has_many :through are mostly identical. So I think in the "Reference" section we only need to mention the methods for has_many and mention they are similar for has_and_belongs_to_many and has_many :through (this one already seems deleted in the PR).

I've removed the "References" sections and added the methods for each association into the "Types of Associations" sections. However, the duplication under those sections for the methods is not ideal. However since we're moving them under the associations, it feels strange to extract them. We also don't have API docs linked to `has_one :through` and `has_many :through` which looks odd.

Let me know what you think of the changes. I'll create a community PR  if you feel its good enough for now? and we can continue this discussion there and maybe also get the communities thoughts.